File: go/string/76_min_window.go
package string

import "math"

//  minWindow 最小覆盖子串
func minWindow(s string, t string) string {
	// 滑动窗口思想
	// need 是需要的字符，win 是窗口中的字符，都用hashMap 记录
	// 当win中的元素满足条件是，win 左边界l 右移 ++ ,不断观察win 中元素是否满足条件

	win, need := make(map[byte]int), make(map[byte]int)

	for _, ch := range t {
		// range 遍历时 ch 是rune 类型，转成byte 类型
		need[byte(ch)]++
	}

	// 窗口长度的值
	min := math.MaxInt32

	// 窗口的左右边界下标
	l, r := 0, 0

	// win 中的元素是否完全匹配到 need 中的个数了
	match := 0
	// 满足条件下的最小长度时的索引
	start, end := 0, 0

	for r < len(s) {

		ch := s[r]
		r++

		// 这个元素是需要的
		if need[ch] > 0 {
			win[ch]++
			if win[ch] == need[ch] {
				match++
			}
		}

		// 窗口中的元素已经完全匹配到，满足条件了
		// match == len(t) 不能用这个，因为t 中可能有重复元素，那match 值 小于 len(t)
		for match == len(need) {
			// 窗口左边界的字符
			chL := s[l]

			// 先记录下满足条件的窗口的左右边界索引
			if r-l < min {
				start, end = l, r
				min = r - l
			}

			// 如果chL 是需要的，则win中的元素个数需要更新
			if need[chL] > 0 {
				if win[chL] == need[chL] {
					match--
				}
				win[chL]--
			}
			l++
		}

	}

	if min == math.MaxInt32 {
		return ""
	}

	return s[start:end]
}

File: rust/src/string/76_min_window.rs
/// 最小覆盖子串
pub fn min_window(s: String, t: String) -> String {
    // 滑动窗口解法
    // win 窗口中的字符, need 是需要的字符，都用 hashMap
    // 当win 满足条件时， 缩小窗口的左边界，l++

    use std::collections::HashMap;

    let (mut win, mut need) = (HashMap::new(), HashMap::new());

    for ch in t.chars() {
        need.entry(ch).and_modify(|x| *x += 1).or_insert(1);
    }

    // 滑动窗口的左右边界
    let (mut l, mut r) = (0, 0);
    // 窗口中已经 满足匹配条件 的字符
    let mut matched = 0;
    // 窗口满足条件时的长度
    let mut min_len = i32::MAX;
    // 窗口满足条件时，s 的索引
    let (mut start, mut end) = (0, 0);

    let s1 = s.chars().collect::<Vec<_>>();

    while r < s1.len() {
        let ch = s1[r];
        r += 1;

        // ch 在need 中 是需要的
        if let Some(cnt) = need.get(&ch) {
            let num = win.entry(ch).and_modify(|x| *x += 1).or_insert(1);

            // 字符 ch 个数满足后，matched ++
            if cnt == num {
                matched += 1;
            }
        }

        // 窗口中的元素满足条件之后
        while matched == need.len() {
            let ch_l = s1[l];

            if r - l < min_len as usize {
                min_len = (r - l) as i32;
                // 记录满足条件的s 的左右索引位置
                println!("win: {:?}, {}", l, r);
                (start, end) = (l, r);
            }

            // 如果 ch_l 在need 中, 判断matched 是否需要减少
            if let Some(&cnt) = need.get(&ch_l) {
                let num = win.get_mut(&ch_l).unwrap();
                if cnt == *num {
                    matched -= 1;
                }
                *num -= 1;
            }
            l += 1;
        }
    }

    if min_len == i32::MAX {
        return "".to_string();
    }
    s[start..end].to_string()
}
